<p>Formatting strings, either with the <code>%</code> operator or <code>str.format</code> method, requires a valid string and arguments matching this
string's replacement fields.</p>
<p>This rule raises an issue when formatting a string will raise an exception because the input string or arguments are invalid. Rule
{rule:python:S3457} covers cases where no exception is raised and the resulting string is simply not formatted properly.</p>
<h2>Noncompliant Code Example</h2>
<pre>
print('Error code %d' % '42')  # Noncompliant. Replace this value with a number as %d requires.

print('User {0} has not been able to access {}'.format('Alice', 'MyFile'))  # Noncompliant. Use only manual or only automatic field numbering, don't mix them.

print('User {a} has not been able to access {b}'.format(a='Alice'))  # Noncompliant. Provide a value for field "b".
</pre>
<h2>Compliant Solution</h2>
<pre>
print('Error code %d' % 42)

print('User {0} has not been able to access {1}'.format('Alice', 'MyFile'))

print('User {a} has not been able to access {b}'.format(a='Alice', b='MyFile'))
</pre>
<h2>See also</h2>
<ul>
  <li> {rule:python:S3457} - Strings should be formatted correctly. </li>
</ul>

