<p>Function, methods and lambdas should not have too many mandatory parameters, i.e. parameters with no default value. Calling them will require code
difficult to read and maintain. To solve this problem you could wrap some parameters in an object, split the function into simpler functions with less
parameters or provide default values for some parameters.</p>
<h2>Noncompliant Code Example</h2>
<p>With a maximum number of 4 parameters:</p>
<pre>
def do_something(param1, param2, param3, param4, param5):  # Noncompliant
	...
</pre>
<h2>Compliant Solution</h2>
<pre>
def do_something(param1, param2, param3, param4):
	...

# Or

def do_something(param1, param2, param3, param4, param5="a default value"):
	...
</pre>
<h2>Exceptions</h2>
<p>The first argument of non-static methods, i.e. <code>self</code> or <code>cls</code>, is not counted as it is mandatory and it is passed
automatically.</p>

