<p>A long parameter list can indicate that a new structure should be created to wrap the numerous parameters or that the function is doing too many
things.</p>
<h2>Noncompliant Code Example</h2>
<p>With a maximum number of 4 parameters:</p>
<pre>
def do_something(param1, param2, param3, param4, param5):
	...
</pre>
<h2>Compliant Solution</h2>
<pre>
def do_something(param1, param2, param3, param4):
	...
</pre>

