<p>The <code>open</code> builtin can open files in different modes, which are provided as a combination of characters. Using an invalid sequence of
characters will at best make <code>open</code> fail, or worse, it will have an undefined behavior (ex: it might ignore some characters).</p>
<p>A valid mode:</p>
<ul>
  <li> should contain one and only one of the following characters: "r" (read), "w" (write), "a" (append), "x" (create). </li>
  <li> should contain zero or one of the following characters: "t" (text), "b" (binary). </li>
  <li> should contain zero or one "+" character (open for updating) </li>
  <li> cannot contain "a", "w", "+", or "x" if "U" (universal newlines) is used. Note that "U" has no effect in python 3, it is deprecated and
  shouldn't be used anymore. </li>
</ul>
<p>This rule raises an issue when an invalid "mode" is provided to the <code>open</code> builtin.</p>
<h2>Noncompliant Code Example</h2>
<pre>
# In python 3 the following fails
# In python 2.7.16 on MacOs, "open" will just ignore the "w" flag
with open("test.txt", "aw") as f:  # Noncompliant
    pass
</pre>
<h2>Compliant Solution</h2>
<pre>
with open("test.txt", "a") as f:
    pass
</pre>
<h2>See</h2>
<ul>
  <li> <a href="https://docs.python.org/3/library/functions.html#open">Python documentation - the <code>open</code> builtin</a> </li>
</ul>

